doc/ui-guidelines.md

```diff
@@ -252,3 +265,150 @@ GOOD:
+
+```clojure
+(defn component [props]
+   (let [some-state (rn/use-state {:name "John Snow"
+                                   :knowledge 0})
+         some-vector [1 2 3]]
+    [rn/view props]))
+```
+
+When the `props` or the bound `"reactive" state` of this component changes, React would re-render this component by executing the function body and by doing so all variables bound to this function's scope are re-initialized meaning they get a new reference if they are reference (non primitive) values. This "re-initialization" and changing of reference happens anytime there is a re-render even if the actual value(s) did not change between renders. This means that any non primitive value in the scope of the function component would by default have an `unstable reference` between every render cycle.
+
+By having an unstable reference, `Object.is` would always return `false` for reference values when comparing them between cycles. The implication of this is that when you pass a reference value with an unstable reference as a prop to a component that is memoized via `React.memo` (without a custom compare function), React would always see that prop as new value when comparing hence re-rendering the component because the reference to the prop changed even when the value hasn't essentially changed. The same is also true for `React.useMemo` and `React.useCallback`, if an unstable reference is passed as a dependency to them, the function they would recompute their function body and return a new reference to the memoized `value` or `callback`. By passing an unstable value (reference value) to these util functions, we essentially defeat the purpose of memoization in the first place.
+
+### Strategies for Creating Stable References
+
+How then do we create stable references? Well there are a number of ways, and there is no one size fits all way of doing it. But here are a few ways we could do it:
+#### Using global variables: 
+We can get stable references by declaring variables in a more global scope relative to the function's local scope. This essentially means if we can, we should declare the variables outside the functions scope. e.g:
+  ```clojure
+  (def some-map {:name "John Snow"
+                :knowledge 0})
+
+  (defn comp []
+    [rn/view {:some-map some-map}])
+  ```
+
+- Trade-offs: While global variables ensure reference stability, they can introduce side effects or global state management complexities. Overuse can make components harder to understand and test due to implicit dependencies.
+
+- Best Scenarios: Use for constants or configuration data that truly does not change over the application's lifetime and does not belong to any component's state.
+  
+#### Using `React.useRef`: 
+Provides a mutable ref object that remains constant throughout the component's lifecycle. Ideal for holding onto a value that does not trigger re-renders.. e.g
+  ```clojure
+  (defn comp []
+    (let [ref (rn/use-ref {:name "John Snow"
```

Comment:
`use-ref` shouldn't be used, we need to remove it if we have it in the code, we have `use-ref-atom` instead

Reply:
I wonder should we consider working with react ref's instead. It might be better to keep things closer to React.

we could do something like:
```
(defn use-ref [default-value]
  (let [ref (rn/use-ref default-value) ;; actual react ref method
        set-ref #(set! (.-current ref) %)]
   [ref set-ref]
  ))
  ```
and the idea then is we have an api similar to `use-state`
e.g
```
(let [[button-ref set-button-ref] (use-ref nil)]
  ...
  )
```
Has `use-ref-atom` worked well so far?

Reply:
yes we considered this, but tbh `use-ref-atom` is closer to react ref than `button-ref set-button-ref` which looks more like a state and even more confusing, `use-ref-atom` works perfect

Reply:
> Has use-ref-atom worked well so far?

`use-ref-atom` has created an issue while compiling for release, the PR addressing it:

https://github.com/status-im/status-mobile/pull/19292#issue-2189606363

> yes we considered this, but tbh use-ref-atom is closer to react ref than button-ref set-button-ref which looks more like a state and even more confusing, use-ref-atom works perfectly

I don't think `use-ref-atom` is closer to react since we must `deref` the returned atom, also we must manually, `swap!` or `reset!` it, IMO the closes to react  is  returning its a value and its setter.

btw a reason to drop ratoms was that people forget to `deref` them, if we provide an atom here, they might also forget something.

Reply:
> use-ref shouldn't be used

Why? it works well.

It's ok to create custom abstractions when needed, but IMO we are over abstracting stuff, e.g.  the `use-mount` is already covered by a proper use of `use-effect`.

Reply:
> e.g. the `use-mount` is already covered

`use-mount` and ``use-unmount` was introduced because of two reasons, 1 - we want to reduce usage of use-effect, because it 99% it is used only for replacing did-mount and will-unmount lifecycle handlers
2- it's easy to read, devs don't need to spend time trying to understand what use-effect is doing, if its for mount or unmount or something else

Reply:
> latest one looks like a state, not as a ref, devs must understand the difference,

I think that's the cleanest, and just because it returns a vector with the value and the setter I don't think looks like a `useState`. We could also return a map but it'd be more verbose.

I'm just worried that the current approach of `use-ref` creates the same problems you tried to avoid with the `ratoms`, such as people forgetting to deref them. Also it's a mix of hook + atom, so may lead to some confusion too.


Anyway, it seems it's an already taken decision, just wanted to contribute with my opinion here.